File: 57-insert-interval/insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<int> merge = newInterval;
        vector<vector<int>> results;
        if (intervals.empty()) {
            return {newInterval};
        }
        for (int i = 0; i < intervals.size(); i++) {
            int curr_start = intervals[i][0];
            int curr_end = intervals[i][1];
            int merge_start = merge[0];
            int merge_end = merge[1];

            if (curr_end < merge_start) { results.push_back(intervals[i]);} 
            else if (merge_end < curr_start) {
                results.push_back(merge);
                results.push_back(intervals[i]);
                for (int j = i + 1; j < intervals.size(); j++) {
                    results.push_back(intervals[j]);
                }
                merge = {};
                break;
            } else {
                merge[0] = std::min(curr_start, merge_start);
                merge[1] = std::max(curr_end, merge_end);
            }
        }
        if (!merge.empty()) { results.push_back(merge);}
        return results;
    }
};

File: 57-insert-interval/insert-interval.py
def is_overlap(a, b):
    return a[0] <= b[1] and b[0] <= a[1]
def merge_intervals(a, b):
    new_start = min(a[0], b[0])
    new_end = max(a[1], b[1])
    return [new_start, new_end]

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[List[int]]
        :type newInterval: List[int]
        :rtype: List[List[int]]
        """

        merged_intervals = []
        has_appended = False
        i = 0
        for i, interval in enumerate(intervals):
            if interval[0] > newInterval[1]:
                merged_intervals.append(newInterval)
                merged_intervals.append(interval)
                has_appended = True
                break
            if is_overlap(interval, newInterval):
                newInterval = merge_intervals(interval, newInterval)
            else:
                merged_intervals.append(interval)
        
        if not has_appended:
            merged_intervals.append(newInterval)

        for j in range(i+1, len(intervals)):
            merged_intervals.append(intervals[j])
        
        return merged_intervals


        